class/practice2.2/main.cpp

```diff
@@ -0,0 +1,125 @@
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <sstream>
+#include <algorithm>
+
+using namespace std;
+
+struct Pixel {
+    int r, g, b;
+};
+
+bool isValidPixel(const Pixel &pixel) {
+    return (pixel.r >= 0 && pixel.r <= 255) && (pixel.g >= 0 && pixel.g <= 255) && (pixel.b >= 0 && pixel.b <= 255);
+}
+
+bool loadPixels(const string &filename, vector<vector<Pixel>> &pixels, string &errorMessage) {
+    ifstream file(filename);
+    if (!file.is_open()) {
+        errorMessage = "Error: Could not open file " + filename;
+        return false;
+    }
+
+    string line;
+    int lineCount = 0;
+    while (getline(file, line)) {
+        if (line.empty()) continue; 
+        stringstream ss(line);
+        vector<Pixel> row;
+        string pixelData;
+        while (getline(ss, pixelData, ' ')) {
+            Pixel pixel;
+            replace(pixelData.begin(), pixelData.end(), ',', ' ');
+            stringstream pixelStream(pixelData);
+            if (!(pixelStream >> pixel.r >> pixel.g >> pixel.b) || !isValidPixel(pixel)) {
+                errorMessage = "Invalid pixel data: " + pixelData;
+                return false;
+            }
+            row.push_back(pixel);
+        }
+        if (row.size() != 16) {
+            errorMessage = "Invalid data format: Each line must contain exactly 16 pixels.";
+            return false;
+        }
+        pixels.push_back(row);
+        lineCount++;
+    }
+
+    if (lineCount != 16) {
+        errorMessage = "Invalid data format: File must contain exactly 16 lines.";
+        return false;
+    }
+
+    return true;
+}
+
+void savePixels(const string &filename, const vector<vector<Pixel>> &pixels) {
+    ofstream file(filename);
+    if (!file.is_open()) {
+        cerr << "Error: Could not open file " << filename << endl;
+        return;
+    }
+
+    for (const auto &row : pixels) {
+        for (size_t i = 0; i < row.size(); ++i) {
+            file << row[i].r << "," << row[i].g << "," << row[i].b;
```

Comment:
you can use std::println to print data to any stream for example
std::print(file, "{},{},{}", row[i].g, row[i].g, row[i].b);
also to avoid usage of indexes we can use join_with function from ranges library
```
for (const auto& row : pixels) {
   std::println(file, row
      | std::ranges::transform([](const auto& pixel) { std::format("{},{},{}", pixel.r, pixel.g, pixel.b);})
      | std::views::join_with(" ")
      | std::ranges::to<std::string>()
   );
}
```
In this code we take a row and transform each pixel to std::string then we join them with space and at the last step all the sequence of chars is converted to std::striing